### ISSUE TITLE
RFC: app config implementation

### ISSUE THREAD

**Issue by mcous** [Opentrons/opentrons#1463]
## overview

This is an RFC ticket for the actual implementation of app settings and config before I start writing any code. Please let me know what you think!

## storage / override mechanisms

How we will store and override the configuration

- Stored in a JSON file in the user's standard app data directory
    - OS specific, same place as logs
    - https://github.com/sindresorhus/electron-store
- Config is applied in the following order:
    1. Use value from CLI arguments
    2. If not specified in CLI, use value from environment variable
    3. If not specified environment, use config file
    4. If not specified in config file, use defaults from code
        - `electron-store` will also write that default value to the file, so updates to the defaults will not change existing installs
    - **Edit**: The order of CLI and environment has been updated to match what `npm` and other JS tools do
- Config file format:
    ```json
    {
      "some": {
        "nested": {
          "value": "foo"
        }
      }
    }
    ```
- Command line argument format:
    ```shell
    /path/to/Opentrons --some.nested.value=foo
    ```
- Environment variable format:
    ```shell
    # posix example but windows would work, too
    export ot_some_nested_value=foo
    /path/to/Opentrons
    ```
- ~Can probably rely on `JSON.parse` to map env var and CLI arg string values to numbers and booleans~
    - ~If we can't there's almost certainly an npm library for this~
    - **Edit**: `yargs-parser` parses numbers automatically, and for booleans we can use either its built in argument negation feature or an empty string
        - e.g. `./Opentrons --disable_features.something` or `OT_FEATURES__SOMETHING='' ./Opentrons`

### feature flags

To enable internal (or possibly external) feature flags in the app, the config will have a `features` section.

### referencing a config value

Due to library needs, we'll probably exclusively reference configs by path:

```js
// just an example, actual config paths TBD
const someFeatureOn = getConfig('features.someFeature.enabled')
```

## main <> renderer

The app has two threads, "main" (the Node thread, where system interaction happens) and "renderer" (the Chrome thread, where React and Redux live). Since "main" handles the config file and overrides, we need a way to get the config from the "main" to the "renderer".

The "renderer" thread will put config in the Redux store. Since UI and actions will be driven by the config, putting it outside the main application state doesn't really make sense.

- At boot, the renderer process will ask for the config via _sync_ RPC
    - Synchronous RPC will be "slow", but it only happens at boot
- In main process, getting config will:
    1. Check overrides for requested path
    2. If no overrides, get requested path from config file
- In renderer process, getting config will follow the same logic, but from state
    - Using selectors, so the override logic from above will probably be duplicated
          - It'll be pretty simple, but we can pull it into a shared library if it gets at all messy
    - Getting them via Electron RPC every time would kill performance

### data alignment

Because config is stored separately in two processes, we need to keep them aligned. Because inter-thread communications happen very quickly (if we keep them async), we can simply only update the Redux store in reaction (eh??) to change events posted by the main thread:

- Config middleware in renderer does initial init mentioned above
- If the UI needs to change a config, it dispatches a `config:SET` action
    1. Middleware captures and checks if value is overridden
        • If overriden, action is a noop
    2. If _not_ overridden, sends _async_ IPC call to main to set value
    3. Main sets value in JSON and `onDidChange` callback is fired
    4. `onDidChange` sends async IPC message to renderer
    5. Config middleware captures message and dispatches `config:UPDATE`
    6. Config reducer updates config in state
- *Important detail*: this implemenation means that users are unable to modify values in the app that are set by CLI args and env vars

**Comment by IanLondon** [comment]
I got some Qs for more context about this:

1. I'm not sure what kind of config variables these are and why they need to change at runtime, could you give a brief couple examples? I'm guessing things like "path to shared-data dir", feature flag toggling... that's all the guesses I got.

2. What's the advantage of nesting config variables, instead of giving them flat names? This might be more obvious when I understand that the config variables are for

3. Are supporting CLI args worth the complexity? Aside from implementation complexity, they add an extra level of hierarchy in this override cascade when you're trying to reason about what config vars are going to your process. When is using CLI args a big advantage over config file + env vars?

**Comment by mcous** [comment]
@IanLondon these are good questions!

> I'm not sure what kind of config variables these are and why they need to change at runtime, could you give a brief couple examples? I'm guessing things like "path to shared-data dir", feature flag toggling... that's all the guesses I got.

So far, what I've got in config is:

- Log level
- Window size
- Entry URL
    - To swap during dev by specifiy `localhost:8090` instead of relying on `NODE_ENV`
    - See <https://12factor.net/config> for why dev environments are less desirable than specific variables

> What's the advantage of nesting config variables, instead of giving them flat names? This might be more obvious when I understand that the config variables are for

Purely organizational. It makes the config file easier to grok by humans, and makes it easier for a single module to pull out an object of only the config it cares about. For example, the `ui` module in `app-shell` will likely do `const config = getConfig('ui')` where `config` is an object of UI configuration.

Example config file:

```json
{
	"log": {
		"level": {
			"file": "debug",
			"console": "info"
		}
	},
	"ui": {
		"width": 1024,
		"height": 768,
		"url": {
			"protocol": "file:",
			"path": "ui/index.html"
		}
	},
	"features": {}
}
```

> Are supporting CLI args worth the complexity? Aside from implementation complexity, they add an extra level of hierarchy in this override cascade when you're trying to reason about what config vars are going to your process. When is using CLI args a big advantage over config file + env vars?

I think yes, they are. CLI args are way easier for our end-users to deal with  (especially Windows users) than environment variables. Also, [`yargs-parser`](https://github.com/yargs/yargs-parser) takes care of mapping `--foo.bar.baz=qux` into `{foo: {bar: {baz: 'qux'}}}`, which removes much of the complexity

>> ...when you're trying to reason about what config vars are going to your process

Super important note. I think with proper logging we can address any potential confusion here, though. Also, by keeping both the config file and the overrides separate in the redux state, we should be able to inspect exactly what the app is using

**Comment by btmorr** [comment]
🦅

**Comment by IanLondon** [comment]
Awesome! That all sounds good.

Might wanna prefix run-app-only env vars with `OT_RUN_APP_` instead of only `OT_` so different Opentrons projects don't collide.

Also, I'm a big advocate of documenting all env vars in some README file; they're slippery.

**Comment by mcous** [comment]
Implemented in #1466 and #1510. Differences in implementation vs RFC:

- Switched the meaning of `config:SET` and `config:UPDATE` to:
    - `UPDATE` - renderer requests main to update
    - `SET` - main informs renderer that update did happen
- Redux state has no knowledge of overrides
    - Complicated implementation and logs provide enough visibility into config
- Config module has no middleware
    - Instead, generic `dispatch` middleware was added to `shell` module
    - Now, main and renderer can dispatch actions to each other

**Comment by IanLondon** [comment]
:fireworks: :fireworks: :fireworks:
